### PR TITLE
DEV/BENCH: Pixi tasks for array backend benchmarks

### PR DESCRIPTION
**Pull request by crusaderky**
- Closes https://github.com/scipy/scipy/issues/24410
- Add documentation on how to run benchmarks with spin
- Add `spin bench --no-build` parameter
- Add pixi tasks:
```
        bench-cpu
        bench-cupy
        bench-dask
        bench-jax
        bench-jax-cuda
        bench-strict
        bench-torch
        bench-torch-cuda
```

# Demo
```
$ pixi r bench-cpu --quick --submodule cluster.Whiten

[50.00%] ··· cluster.Whiten.time_whiten                                                                                                                 ok
[50.00%] ··· ================== ========== ============
             --                          shape         
             ------------------ -----------------------
                  backend        (10, 10)   (100, 100) 
             ================== ========== ============
                   numpy         174±0μs     212±0μs   
              array_api_strict   363±0μs     384±0μs   
                    cupy           n/a         n/a     
                 torch:cpu       1.30±0ms    1.35±0ms  
                 torch:cuda        n/a         n/a     
                 dask.array      5.19±0ms    5.16±0ms  
               jax.numpy:cpu     31.6±0μs    99.8±0μs  
               jax.numpy:cuda      n/a         n/a     
             ================== ========== ============
```

CC @lucascolley

### CODE REVIEW

**Comment by crusaderky** [comment]
I originally added an environment for all CUDA backends, which would result in no n/a in the table above, but that would have involved either
- downgrading cupy from cuda 13 to cuda 12 (I cannot reproduce the issues mentioned in #23458)
- upgrading torch and jax from cuda 12 to cuda 13. This in turn implies moving both packages from conda to pip and bumping up the JAX version by 6 months.

both would be firmly out of scope for this PR.

File: .spin/cmds.py [hunk 1]
```diff
@@ -808,6 +808,9 @@ def _dirty_git_working_dir():
     required=False,
     nargs=-1
 )
+@click.option(
+    '--no-build', default=False, is_flag=True,
+    help='Build SciPy before running benchmarks')
```
Comment (lucascolley) [inline .spin/cmds.py]: we can use `@meson.build_option` instead, see https://github.com/scipy/scipy/pull/24336

**Review by lucascolley — CHANGES_REQUESTED** [review]
thanks!

File: benchmarks/README.rst [hunk 2]
```diff
@@ -40,6 +40,23 @@ Command-line help is available as usual via ``asv --help`` and
...
+
+  spin bench -b all
+
+Unlike `spin test`, `spin bench` disregards the ``SCIPY_DEVICE`` flag. Instead, if a
+CUDA device is installed for backends that support both CPU and GPU (e.g. JAX, PyTorch),
+the benchmark will be run on both CPU and GPU and the results will be shown side by
+side.
+
+The JAX backend is compiled with `jax.jit` when supported; JIT warmup times are excluded
+from the benchmark results.
```
Comment (lucascolley) [inline benchmarks/README.rst:57]: does this require something like https://github.com/scipy/scipy/blob/99744dd3dd29218a1fc8cf2098ceece309a0edb2/benchmarks/benchmarks/cluster.py#L28-L29 ? If so can we mention that?

**Review by lucascolley — COMMENTED** [review]
all looks great otherwise

**Review by lucascolley — COMMENTED** [review]
not blocking for this PR but there is also https://github.com/scipy/scipy/blob/main/doc/source/dev/contributor/benchmarking.rst

File: benchmarks/README.rst (hunk 2 above)
Comment (lucascolley) [reply benchmarks/README.rst:57]: > does this require something like {`if self.warmup`}

gentle ping @crusaderky

Comment (crusaderky) [reply benchmarks/README.rst:57]: Added pointer to the docstring of `common.XPBenchmark` for further documentation

**Review by lucascolley — APPROVED** [review]
thanks Guido!

**Comment by crusaderky** [comment]
> not blocking for this PR but there is also https://github.com/scipy/scipy/blob/main/doc/source/dev/contributor/benchmarking.rst

Updated
